flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -2838,6 +2843,37 @@ mlir::Value IntrinsicLibrary::genSign(mlir::Type resultType,
   return builder.create<mlir::SelectOp>(loc, cmp, neg, abs);
 }
 
+// SIZE
+fir::ExtendedValue
+IntrinsicLibrary::genSize(mlir::Type resultType,
+                          llvm::ArrayRef<fir::ExtendedValue> args) {
+  // TODO: handle assumed-rank arrays, especially a dummy whose actual argument
+  // is an assumed-size array
+  assert(args.size() == 3);
+
+  // Calls to SIZE that don't have the DIM argument are handled elsewhere
+  assert(!isAbsent(args[1]));
+
+  // Handle the ARRAY argument
+  auto array = builder.createBox(loc, args[0]);
+
+  // Handle the optional DIM argument.  Convert the Fortran 1-based index to
+  // the FIR 0-based index
+  auto indexType = builder.getIndexType();
+  auto oneBasedDim =
+      builder.createConvert(loc, indexType, fir::getBase(args[1]));
+  auto one = builder.createIntegerConstant(loc, indexType, 1);
+  auto zeroBasedDim = builder.create<mlir::SubIOp>(loc, oneBasedDim, one);
```

Comment:
Can `DIM` ever be anything other than 1-based in this context?

Reply:
No.  Here's what the standard says:
    DIM (optional) shall be an integer scalar with a value in the range
      1<=DIM<=n, where n is the rank of ARRAY.